### PR TITLE
v5 refactor and less optimisation

### PR DESCRIPTION
**Pull request by guywillis**

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -1,23 +1,24 @@
...
+  "issues": "https://github.com/cgkineo/adapt-mediaWithBenefits/issues/ ",
   "component": "mediawithbenefits",
+  "displayName": "Media with benefits",
+  "description": "Extends the functionality of adapt-contrib-media by providing a gallery of videos",
+  "main": "/js/adapt-mediaWithBenefits.js",
   "assetFields" : [
     "_media.ogv",
     "_media.mp4",
     "_media.poster"
   ],
```
Comment (moloko) [inline bower.json]: You can remove `assetFields`. We [took it out of the media component back in 2019](https://github.com/adaptlearning/adapt-contrib-media/pull/202)

File: bower.json [hunk 2]
```diff
@@ -1,23 +1,24 @@
 {
   "name": "adapt-mediaWithBenefits",
-  "version": "0.0.1",
-  "framework": ">=3",
+  "version": "0.0.2",
```
Comment (moloko) [inline bower.json]: We've done major version bumps for all the other v5 compatible releases
```suggestion
  "version": "1.0.0",
```

File: example.json [hunk 3]
```diff
@@ -1,129 +1,123 @@
...
-                "inlineTranscriptBody": "Transcript body",
-                "transcriptLinkButton": "Transcript",
-                "transcriptLink": "assets/transcript.pdf"
-            },
-            "_graphic": {
-                "src": "course/en/video/video2_thumb.png"
-            },
-            "title": "Video 2",
-            "ariaLabel": "Select this to hear from Speaker 2"
+        "_playerOptions": {
```
Comment (moloko) [inline example.json:79]: might be easier just to leave this out and point people at the example.json for the media component

File: js/adapt-mediaWithBenefits.js [hunk 4]
```diff
@@ -1,184 +1,185 @@
...
-            }, Media.view.prototype.events);
-        },
-
-        onMediaElementPlay: function() {
-            Media.view.prototype.onMediaElementPlay.apply(this, arguments);
+  var MediaWithBenefitsModel = ComponentModel.extend({
+    init:function() {
+      this.set('_media', this.get('_items')[0]._media);
+      this.set('_originalTranscript', this.get('_transcript'));
+    }
```
Comment (moloko) [inline js/adapt-mediaWithBenefits.js]:
```suggestion
    init: function() {
      this.set({
        _media: this.get('_items')[0]._media,
        _originalTranscript: this.get('_transcript')
      });
    }
```

File: js/adapt-mediaWithBenefits.js [hunk 5]
```diff
@@ -1,184 +1,185 @@
...
+      var str = Media.view.prototype.className.apply(this, arguments);
+      str += ' media__component';
+      return str;
+    },
 
-            this.markItemAsWatched();
-        },
+    events: function() {
+      return _.extend({
+        "click .js-mediawithbenefits-item": "onItemClicked"
```
Comment (moloko) [inline js/adapt-mediaWithBenefits.js]:
```suggestion
        'click .js-mediawithbenefits-item': 'onItemClicked'
```

File: bower.json [hunk 6]
```diff
@@ -1,23 +1,24 @@
...
+  "framework": ">=5",
```
Comment (moloko) [inline bower.json]: Due to use of ES6 arrow functions
```suggestion
  "framework": ">=5.3",
```

File: js/adapt-mediaWithBenefits.js [hunk 7]
```diff
@@ -1,184 +1,185 @@
...
+    selectItem:function(index) {
+      // get the selected item configuration
+      var itemCfg = this.model.get('_items')[index];
+      // get the selected element
+      var $selectedItem = this.$('.js-mediawithbenefits-item').eq(index);
+
+      // update classes
+      this.$('.js-mediawithbenefits-item').removeClass('is-selected');
+      $selectedItem.addClass('is-selected');
+      $selectedItem.addClass('is-visited');
```
Comment (moloko) [inline js/adapt-mediaWithBenefits.js]:
```suggestion
      $selectedItem.addClass('is-selected is-visited');
```

File: js/adapt-mediaWithBenefits.js [hunk 8]
```diff
@@ -1,184 +1,185 @@
...
+      } else {
+        $transcriptBodyContainer.stop(true,true).slideDown(function() {
+          $(window).resize();
+        });
+        $button.attr('aria-expanded', true);
+        $transcriptBodyContainer.addClass("inline-transcript-open");
+        $buttonText.html(this.model.get("_transcript").inlineTranscriptCloseButton);
+
+        this.checkCompletionByTranscript();
+      }
```
Comment (moloko) [inline js/adapt-mediaWithBenefits.js]:
```suggestion
      var $transcriptBodyContainer = this.$('.media__transcript-body-inline');
      var $button = this.$('.media__transcript-btn-inline');
      var $buttonText = $button.find('.media__transcript-btn-text');

      if ($transcriptBodyContainer.hasClass('inline-transcript-open')) {
        $transcriptBodyContainer.stop(true, true).slideUp(function() {
          $(window).resize();
        }).removeClass('inline-transcript-open');
        $button.attr('aria-expanded', false);
        $buttonText.html(this.model.get('_transcript').inlineTranscriptButton);
        return;
      }
      
      $transcriptBodyContainer.stop(true, true).slideDown(function() {
        $(window).resize();
      }).addClass('inline-transcript-open');
      $button.attr('aria-expanded', true);
      $buttonText.html(this.model.get('_transcript').inlineTranscriptCloseButton);

      this.checkCompletionByTranscript();
```

File: js/adapt-mediaWithBenefits.js [hunk 9]
```diff
@@ -1,184 +1,185 @@
...
+    },
+
+    onExternalTranscriptClicked: function(event) {
+      this.markItemAsWatched();
+      this.checkCompletionByTranscript();
+    },
+  }, {
+    template: "mediawithbenefits"
```
Comment (moloko) [inline js/adapt-mediaWithBenefits.js]:
```suggestion
    template: 'mediawithbenefits'
```

File: js/adapt-mediaWithBenefits.js [hunk 10]
```diff
@@ -1,184 +1,185 @@
...
+  });
+
+  return Adapt.register('mediawithbenefits', {
+    model:MediaWithBenefitsModel,
+    view:MediaWithBenefitsView
+  });
```
Comment (moloko) [inline js/adapt-mediaWithBenefits.js]:
```suggestion
  return Adapt.register('mediawithbenefits', {
    model: MediaWithBenefitsModel,
    view: MediaWithBenefitsView
  });
```

**Review by moloko — COMMENTED** [review]
few suggestions following code 👓 

I think the transcript stuff in example.json could use simplifying - I think it's only necessary to attach properties you actually want to change to each of the `_items` but I could do with actually testing all that rather than trying to do it purely by reading the code!

**Comment by moloko** [comment]
Just to note we should probably do a tag/release of the v3/4 compatible version before merging this

File: less/mediawithbenefits.less [hunk 11]
```diff
@@ -1,72 +1,57 @@
...
+  &__item-image {
+    .transition(all .2s ease);
+  }
+}
+
+// THEME
+// --------------------------------------------------
+.mediawithbenefits {
+  &__item-container {
+    text-align: center;
```
Comment (kirsty-hames) [inline less/mediawithbenefits.less]: I don't think `text-align` does anything here so this could be removed?

File: less/mediawithbenefits.less [hunk 12]
```diff
@@ -1,72 +1,57 @@
...
+  }
+
+  &__item-title {
+    margin-top: @component-title-margin / 2;
+    .item-title;
+    text-align: center;
```
Comment (kirsty-hames) [inline less/mediawithbenefits.less]: `text-align` isn't needed here as this is inherited from the button

File: less/mediawithbenefits.less [hunk 13]
```diff
@@ -1,72 +1,57 @@
...
+  }
+
+  &__item-subtitle {
+    margin-top: @component-body-margin / 2;
+    text-align: center;
```
Comment (kirsty-hames) [inline less/mediawithbenefits.less]: `text-align` isn't needed here as this is inherited from the button

File: less/mediawithbenefits.less [hunk 14]
```diff
@@ -1,72 +1,57 @@
...
+  }
+
+  &__item-image {
+    border: 3px solid @item-color;
+  }
+
+  .no-touch &__item:hover &__item-image {
```
Comment (kirsty-hames) [inline less/mediawithbenefits.less]: For consistency with other hover styles could we add a `:not` condition for `.is-selected` please?

`.no-touch &__item:not(.is-selected):hover &__item-image`

File: less/mediawithbenefits.less [hunk 15]
```diff
@@ -1,72 +1,57 @@
...
+    border-color: @item-color-hover;
+  }
+
+  &__item.is-visited &__item-image {
+    border-color: @visited;
+  }
+}
```
Comment (kirsty-hames) [inline less/mediawithbenefits.less:55]: Can we add a selected border style please?

```
  &__item.is-selected &__item-image {
    border-color: @item-color-selected;
  }
```

**Review by kirsty-hames — COMMENTED** [review]
Looks good thanks @guywillis. I've just added some comments around the theme less.

File: js/adapt-mediaWithBenefits.js (hunk 8 above)
Comment (guywillis) [reply js/adapt-mediaWithBenefits.js]: I have edited the js as per the suggestion above but left the if else statement in it's current form as changing it caused the closed transcript button text not to appear.

File: example.json [hunk 16]
```diff
@@ -57,6 +52,7 @@
                     ]
                 },
                 "_transcript": {
+                    "_comment._inlineTranscript": "Enable item transcript with overrides",
```
Comment (kirsty-hames) [inline example.json]: Not an issue for this component as it's not AAT supported but just FYI the use of '.' is invalid json (for reference I raised the issue on List https://github.com/cgkineo/adapt-list/issues/23).

Perhaps for consistency it's worth using the generic key suggested in the issue?

"_comment": "inlineTranscript = Enable item transcript with overrides",

**Review by kirsty-hames — APPROVED** [review]
Looks good thanks

**Comment by moloko** [comment]
@oliverfoster I'm guessing we would prefer the model to be in separate .js file?

**Comment by moloko** [comment]
One issue I've noticed in testing this is that if you set `_playerOptions` they don't seem to get applied... not managed to figure out why yet, will investigate further next week.

**Comment by moloko** [comment]
> 
> 
> One issue I've noticed in testing this is that if you set `_playerOptions` they don't seem to get applied... not managed to figure out why yet, will investigate further next week.

(Just to say I don't think this is anything to do with the v5 upgrade changes, suspect it has always been an issue)

**Comment by oliverfoster** [comment]
> @oliverfoster I'm guessing we would prefer the model to be in separate .js file?

Yea, one file for the model, one for the view and one for the register. [Slider is a good example](https://github.com/adaptlearning/adapt-contrib-slider/tree/master/js).

**Comment by chris-steele** [comment]
Does anyone feel a name change would be in order at this grand stage?

**Review by moloko — APPROVED** [review]
looks good to me, tested in FF, Chrome and IE11

there's an issue with some of the `_playerOptions` not being applied but I seriously doubt it was caused by any of these changes - will log it separately.

**Comment by moloko** [comment]
> Does anyone feel a name change would be in order at this grand stage?

adapt-mediaGallery? Or just adapt-videoGallery given that it seems quite unlikely anyone would ever want to use this with just audio (although, now I've said that you can bet someone will)

**Comment by chris-steele** [comment]
`adapt-mediaGallery` was my thought

**Pull request merged**